Sources/LanguageServerProtocol/Command.swift

```diff
@@ -0,0 +1,176 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SKSupport
+
+/// Represents a reference to a command identified by a string. Used as the result of
+/// requests that returns actions to the user, later used as the parameter of
+/// workspace/executeCommand if the user wishes to execute said command.
+public struct Command: Codable, Hashable {
+
+  /// The title of this command.
+  public var title: String
+
+  /// The internal identifier of this command.
+  public var command: String
+
+  /// The arguments related to this command.
+  public var arguments: [AnyCommandArgument]?
+
+  public init(title: String, command: String, arguments: [AnyCommandArgument]?) {
+    self.title = title
+    self.command = command
+    self.arguments = arguments
+  }
+}
+
+public protocol CommandArgumentType: Codable, Hashable {}
```

Comment:
Could we use an `enum` for this instead?  There are a few places in this patch where we make assumptions about the set of types that can conform to this protocol, so this may be better as an enum where we will get an error if we miss a case.  Splitting Int/Double is a bit weird, but may be unavoidable.

```swift
enum CommandArgumentType {
  case null // JSON is great, isn't it?
  case int(Int)
  case double(Double)
  case bool(Bool)
  case string(String)
  case array([CommandArgumentType])
  case dictionary([String: CommandArgumentType])
}
```